### DISCUSSION TITLE
Can Z3 answer that question?

### DISCUSSION THREAD

**Discussion by 14H034160212** [Z3Prover/z3#5652] [Q&A]
Hi,

I am trying to use SAT Solver to answer the following question. The context is given by (¬α->¬β,¬β->¬γ). For each option, there are ( ¬ γ -> ¬ α ), ( γ -> α ), ( ¬ γ -> ¬ β ), and ( α -> γ ). Option B can be selected using contraposition law and transitive law. Can Z3 select the correct answer by given the context? If yes, how to do that? If not, is there any other SAT solver can do that? Many thanks.
<img width="282" alt="5729cfe1ced321b1166213499a2ad16" src="https://user-images.githubusercontent.com/23516191/140310492-14ad8eeb-d938-4730-8a51-e2e269bfe0ab.png">

**Comment by aytey** [comment]
I think it seems quite unclear what you're trying to; do you want Z3 to tell you the possible options out of the context?

If so, it seems that you want to assert `¬a -> ¬b` and `¬b -> ¬c` and then use Z3 to find all solutions to model (which can be done using https://theory.stanford.edu/~nikolaj/programmingz3.html#sec-blocking-evaluations)

**Reply by 14H034160212** [reply]
Thanks for your reply! Yes, I think you understand that. In more specifically, I want to find as many as possible solutions which are logically equivalent to the context rules (¬a -> ¬b, ¬b -> ¬c) and the logical extension of those two context rules using some logical rules like contraposition law, transitivity law, and so on. I try to use the code from the link and assert ¬a -> ¬b and ¬b -> ¬c, but it gives me an empty list. Is it correct? 
![image](https://user-images.githubusercontent.com/23516191/140416392-f6c691ce-d8fe-46b2-a5c6-96a4e511f86f.png)

**Reply by aytey** [reply]
There's at least two things wrong here: 

1. You never `s.assert(terms)`
2. You need to block `s.model()` not `terms`

**Reply by 14H034160212** [reply]
Thanks a lot for your suggestion! I got some results after fixing the above issues. 
![image](https://user-images.githubusercontent.com/23516191/140461588-cf0c6fd8-3d99-4474-b04a-b8ba5740c00d.png)

**Reply by 14H034160212** [reply]
But it seems Z3 cannot know which option can be derived from the context rules. For example, Z3 can find a set of solutions for each variable to make the formula satisfable when I add option A into the context rules. Do you have any idea to avoid that happening? In this example, I hope only option B is correct, which is the only option that can be derived from context rules.
![image](https://user-images.githubusercontent.com/23516191/140462271-7ba11ab1-e0f8-4822-a044-4776f6bbc26e.png)

**Reply by aytey** [reply]
Asserting `A -> B` *does not* mean that `A` should be true (False -> anything == True)

**Reply by 14H034160212** [reply]
I see. You got the point. So, I guess SAT Solver cannot only select option B as the correct option by giving these context rules? Cause even I set all variables as True, I can also derive other options as True, not only option B is correct.

**Reply by aytey** [reply]
The solver here (which may or may not be a SAT solver, depending on your formula) is welcome to make any assignment, meeting the assertions you add. If you don't want "other options" then you need to add a clause that says "and not this".

**Reply by 14H034160212** [reply]
OK. I thought the SAT solver could automatically find those other options that cannot be derived from the context rules.

**Reply by 14H034160212** [reply]
I write the context rules and optionB in the following way, but the return result is an empty list. 
((not a  -> not b) and (not b -> not c)) -> (c -> a)
![image](https://user-images.githubusercontent.com/23516191/140493520-a029826e-2ecc-412a-a839-2beb0c66cb48.png)
What is the meaning of the empty list? Cause I also try another simple example. I also got an empty list.
(a->b and b->c) -> (a->c)
![image](https://user-images.githubusercontent.com/23516191/140499092-1e152e34-ade7-4100-81b1-5879d067b435.png)
If there is no solution, the console will print no solution like that. So, maybe empty list [] means a different meaning?
![image](https://user-images.githubusercontent.com/23516191/140503990-f9fbe46a-9bce-4198-8e21-9652a0ff2f8f.png)

**Reply by 14H034160212** [reply]
I finally understand that the specific use here can be changed to the rule in the given context as a condition, each option is used as a conclusion, and the condition is used to derive the opposite of the conclusion, which shows that the opposite of the conclusion is unsatisfiable, so the conclusion is true. I tried it with this method, and it’s okay. Now only the opposite of optionB is unsat, and the others are sat. It can be seen that only optionB is correct.
The negation of Option A: ((not a  -> not b) and (not b -> not c)) -> not ( not c -> not a)
The negation of Option B:  ((not a  -> not b) and (not b -> not c)) -> not (c -> a)
The negation of Option C: ((not a  -> not b) and (not b -> not c)) -> not ( not c -> not b)
The negation of Option D: ((not a  -> not b) and (not b -> not c)) -> not ( a -> c)
And then only the negation of option B return unsat and the others all return sat . So I can ensure that option B is definitely true.
```
from z3 import *
a = Bool('a')
b = Bool('b')
c = Bool('c')
s = Solver()

s.add(Or(a, (Not(b))))
s.add(Or(b, (Not(c))))
print("case a")
s.push()
s.add(Not(Or(c,Not(a))))
print(s.check())
s.pop()

print("case b")
s.push()
s.add(Not(Or(Not(c),a)))
print(s.check())
s.pop()

print("case c")
s.push()
s.add(Not(Or(c,Not(b))))
print(s.check())
s.pop()

print("case d")
s.push()
s.add(Not(Or(Not(a),c)))
print(s.check())
s.pop()
```
